lib/cosmos/top_level.rb

```diff
@@ -655,4 +656,25 @@ def self.kill_thread(owner, thread, graceful_timeout = 1, timeout_interval = 0.0
     end
   end
 
+  # Close a socket in a manner that ensures that any reads blocked in select
+  # will unblock across platforms
+  # @param socket The socket to close
+  def self.close_socket(socket)
+    if socket
+      # Calling shutdown and then sleep seems to be required
+      # to get select to reliably unblock on linux
+      begin
+        socket.shutdown(:RDWR)
+        sleep(0)
+      rescue Exception
+        # Oh well we tried
+      end
```

Comment:
Just closing a socket seems to unreliably unblock select on linux/mac.   This fix worked well.